pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotLeadControllerRestletResource.java

```diff
@@ -0,0 +1,206 @@
+import org.apache.helix.HelixManager;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.MasterSlaveSMD;
+import org.apache.pinot.common.config.TableNameBuilder;
+import org.apache.pinot.common.utils.CommonConstants.Helix;
+import org.apache.pinot.common.utils.helix.LeadControllerUtils;
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+@Api(tags = Constants.LEAD_CONTROLLER_TAG)
+@Path("/leader")
+public class PinotLeadControllerRestletResource {
+  public static Logger LOGGER = LoggerFactory.getLogger(PinotLeadControllerRestletResource.class);
+
+  @Inject
+  private PinotHelixResourceManager _pinotHelixResourceManager;
+
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  @Path("/tables")
+  @ApiOperation(value = "Gets leaders for all tables", notes = "Gets leaders for all tables")
+  public LeadControllerResponse getLeadersForAllTables() {
+    Map<String, LeadControllerEntry> leadControllerEntryMap = new LinkedHashMap<>();
+    HelixManager helixManager = _pinotHelixResourceManager.getHelixZkManager();
+    boolean isLeadControllerResourceEnabled;
+    try {
+      isLeadControllerResourceEnabled = LeadControllerUtils.isLeadControllerResourceEnabled(helixManager);
+    } catch (Exception e) {
+      throw new ControllerApplicationException(LOGGER,
+          "Exception when checking whether lead controller resource is enabled or not.",
+          Response.Status.INTERNAL_SERVER_ERROR);
+    }
+
+    // Returns empty map if lead controller resource is disabled.
+    if (!isLeadControllerResourceEnabled) {
+      return new LeadControllerResponse(false, leadControllerEntryMap);
+    }
+
+    ExternalView leadControllerResourceExternalView = getLeadControllerResourceExternalView(helixManager);
+    for (int partitionId = 0; partitionId < Helix.NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE; partitionId++) {
+      String partitionName = LeadControllerUtils.generatePartitionName(partitionId);
+      String participantInstanceId =
+          getParticipantInstanceIdFromExternalView(leadControllerResourceExternalView, partitionName);
+      if (participantInstanceId == null) {
+        continue;
+      }
+      leadControllerEntryMap
+          .putIfAbsent(partitionName, new LeadControllerEntry(participantInstanceId, new ArrayList<>()));
+    }
+
+    // Assigns all the tables to the relevant partitions.
+    List<String> tableNames = _pinotHelixResourceManager.getAllTables();
+    for (String tableName : tableNames) {
+      String rawTableName = TableNameBuilder.extractRawTableName(tableName);
+      int partitionId = LeadControllerUtils.getPartitionIdForTable(rawTableName);
+      String partitionName = LeadControllerUtils.generatePartitionName(partitionId);
+      LeadControllerEntry leadControllerEntry = leadControllerEntryMap.get(partitionName);
+      leadControllerEntry.getTableNames().add(tableName);
```

Comment:
Is there a possibility that leadControllerEntry is null (say, due to some helix issue)? Perhaps better to check for null here. Better, we can add the entry here instead of a separate loop above?

Reply:
`putIfAbsent` method is used so there's no way that leadControllerEntry is null.